src/client.ts

```diff
@@ -117,41 +144,117 @@ export class OpenFeatureClient implements Client {
       options: FlagEvaluationOptions | undefined
     ) => Promise<ResolutionDetails<T>>,
     defaultValue: T,
-    context: EvaluationContext = {},
+    flagType: FlagType,
+    invocationContext: EvaluationContext = {},
     options: FlagEvaluationOptions = {}
   ): Promise<EvaluationDetails<T>> {
     // merge global, client, and evaluation context
-    const mergedContext = {
-      ...this.api.context,
+
+    const allHooks = [...OpenFeature.hooks, ...this.hooks, ...(options.hooks || [])];
+    const allHooksReversed = [...allHooks].reverse();
+
+    // merge global and client contexts
+    const globalAndClientContext = {
+      ...OpenFeature.context,
       ...this.context,
-      ...context,
+    };
+
+    // this reference cannot change during the course of evaluation
+    // it may be used as a key in WeakMaps
+    const hookContext: Readonly<HookContext> = {
+      flagKey,
+      defaultValue,
+      flagType,
+      client: this,
+      provider: OpenFeature.provider,
+      context: globalAndClientContext,
     };
 
     try {
-      // if a transformer is defined, run it to prepare the context.
+      const mergedHookContext = await this.beforeHooks(allHooks, hookContext, options);
+
+      // merge context in order: global, client, hook, invocation
+      const mergedContext = {
+        ...mergedHookContext,
+        ...invocationContext,
+      };
+
+      // if a transformer is defined, run it to prepare the context
       const transformedContext =
         typeof this.provider.contextTransformer === 'function'
           ? await this.provider.contextTransformer(mergedContext)
           : mergedContext;
 
       // run the referenced resolver, binding the provider.
       const resolution = await resolver.call(this.provider, flagKey, defaultValue, transformedContext, options);
-      return {
+
+      const evaluationDetails = {
         ...resolution,
         flagKey,
       };
+
+      await this.afterHooks(allHooksReversed, hookContext, evaluationDetails, options);
+
+      return evaluationDetails;
     } catch (err: unknown) {
       const errorCode = (!!err && (err as { code: string }).code) || GENERAL_ERROR;
+
+      await this.errorHooks(allHooksReversed, hookContext, err, options);
+
       return {
         errorCode,
         value: defaultValue,
         reason: ERROR_REASON,
         flagKey,
       };
+    } finally {
+      await this.finallyHooks(allHooksReversed, hookContext, options);
+    }
+  }
+
+  private async beforeHooks(hooks: Hook[], hookContext: HookContext, options: FlagEvaluationOptions) {
+    for (const hook of hooks) {
+      // freeze the hookContext
+      Object.freeze(hookContext);
+
+      // use Object.assign to avoid modification of frozen hookContext
+      Object.assign(hookContext.context, {
+        ...hookContext.context,
+        ...(await hook?.before?.(hookContext, Object.freeze(options.hookHints))),
+      });
+    }
```

Comment:
Normally I don't like doing await in loops, however, I think if we want to support async hooks, we can't just do a `Promise.all` - that might result in unexpected order of evaluation.

@beeme1mr